### PR TITLE
Add scripts to upgrade configuration

### PR DESCRIPTION
**Pull request by John-Holt-Tessella**
See https://github.com/ISISComputingGroup/EPICS-inst_servers/pull/117

### CODE REVIEW

File: src/upgrade_step_from_3p2p1.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return 0
+
+    def _check_final_file_contains_alarm_ioc(self, logger, modified_file_contents):
+        et = ElementTree.fromstringlist(modified_file_contents)
+        ns = {"ioc_ns": 'http://epics.isis.rl.ac.uk/schema/iocs/1.0'}
+        node_count = len(et.findall(r".//ioc_ns:ioc[@name='ALARM']", ns))
+        if node_count != 1:
+            # I can not see how to generate this error but it is here because it is important
+            logger.error("IOC default component file now contains contains {0} ALARM iocs it should contain exactly 1."
+                         .format(node_count))
```
Comment (Tom-Willemsen) [inline src/upgrade_step_from_3p2p1.py]: I got this error when I modified my `iocs.xml` to look like this:

```
<?xml version="1.0" ?>
<iocs xmlns="http://epics.isis.rl.ac.uk/schema/iocs/1.0" xmlns:ioc="http://epics.isis.rl.ac.uk/schema/iocs/1.0" xmlns:xi="http://www.w3.org/2001/XInclude">
    <ioc autostart="true" name="INSTETC_01" restart="true" simlevel="none"><macros/><pvs/><pvsets/></ioc>
    <ioc autostart="true" name="ISISDAE_01" restart="true" simlevel="none"><macros/><pvs/><pvsets/></ioc>	
    <!-- Above this line only change in master branch -->	
</iocs>
```

I feel like it should work with any valid XML that conforms to the schema and not rely on the formatting of the XML. In this case we're relying on the `</ioc>` tag being on it's own line.

File: test/mother.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        <pvsets/>
+    </ioc>
+    <ioc autostart="true" name="ISISDAE_01" restart="true" simlevel="none">
+        <macros/>
+        <pvs/>
+        <pvsets/>
+    </ioc>
+    <!-- Above this line only change in master branch -->
+</iocs>
+""".split("\n")
```
Comment (Tom-Willemsen) [inline test/mother.py]: Should this be `os.linesep` for compatibility? (comment repeats as necessary)

File: src/upgrade_step_from_3p2p1.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        modified_file_contents = []
+        add_after_next_ioc_element_close = False
+        for line in ioc_file:
+            modified_file_contents.append(line)
+            if line.strip().startswith("<ioc") and '"ISISDAE_01"' in line:
+                add_after_next_ioc_element_close = True
+
+            if add_after_next_ioc_element_close and line.strip().startswith("</ioc>"):
+                indent = line.find("</ioc>")
+                for xml_line_to_add in XML_TO_ADD.split("\n"):
```
Comment (Tom-Willemsen) [inline src/upgrade_step_from_3p2p1.py]: `os.linesep` instead of `\n` here?

File: src/local_logger.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+            message: message to write (no new lines needed)
+
+        Returns:
+
+        """
+
+        formatted_message = " INFO: {0}\n".format(message)
+        with file(self._log_file, mode="a") as f:
+                f.write(formatted_message)
+        sys.stdout.write(formatted_message)
```
Comment (John-Holt-Tessella) [inline src/local_logger.py]: Plus won't fail test

File: test/mother.py (hunk 2 above)
Comment (John-Holt-Tessella) [inline test/mother.py]: NB Has to be \n so xml is produced correctly (not sure why)

**Pull request merged**